base_layer/blockchain/src/chain.rs

```diff
@@ -109,40 +67,90 @@ impl Chain {
         result
     }
 
+    /// Internal helper function to do orphan logic
+    fn orphaned_block(&mut self, new_block: Block) -> Result<(), ChainError> {
+        if self.orphans.contains_key(&new_block.header.hash()[..]) {
+            return Err(ChainError::StateProcessingError(StateError::DuplicateBlock));
+        };
+        let mut hash = [0; 32];
+        hash.copy_from_slice(&new_block.header.hash());
+        let pow = new_block.header.pow.clone();
+        self.orphans.insert(hash, new_block);
+        let mut currently_used_orphans: Vec<BlockHash> = Vec::new();
+        if pow > self.current_total_pow {
+            // we have a potential re-org here
+            let result = self.handle_re_org(&hash, 1, &mut currently_used_orphans);
+            let result = if result.is_err() {
+                self.block_chain_state.reset_chain_state()?;
+                result
+            } else {
+                for hash in &currently_used_orphans {
+                    self.orphans.remove(hash);
+                }
+                self.current_total_pow = pow;
+                self.block_chain_state
+                    .save_state()
+                    .map_err(ChainError::StateProcessingError)
+            };
+            result
+        } else {
+            Err(ChainError::StateProcessingError(StateError::OrphanBlock))
+        }
+    }
+
     /// Internal recursive function to handle re orgs
-    fn handle_re_org(&mut self, block_hash: &BlockHash, counter: usize) -> Result<(), ChainError> {
+    fn handle_re_org(
+        &mut self,
+        block_hash: &BlockHash,
+        counter: usize,
+        mut currently_used_orphans: &mut Vec<BlockHash>,
+    ) -> Result<(), ChainError>
+    {
         // The searched hash should always be in the orphan list
-        self.currently_used_orphans.push(block_hash.clone()); // save all orphans we have used
-        let block = self.orphans.get(block_hash).unwrap().clone();
-        if self
-            .block_chain_state
-            .headers
-            .get_object(&block.header.prev_hash)
-            .is_some()
-        {
+        currently_used_orphans.push(block_hash.clone()); // save all orphans we have used
+        let block = self.orphans.get(block_hash).unwrap().clone(); // todo remove this clone
+        let parent = self.block_chain_state.headers.get_object(&block.header.prev_hash);
+        if parent.is_some() {
             // we know of parent so we can re-org
-            self.block_chain_state.rewind_state(counter)?;
+            let h = parent.unwrap().height;
+            self.block_chain_state
+                .rewind_state((self.block_chain_state.get_tip_height() - h) as usize)?;
```

Comment:
Question - We just throw away the former tip chain here from the looks of things. Shouldn't these blocks go back into the orphan pool?

Reply:
It could be an optimisation we could look into.  But as of now we dont actually store the blocks, we store parts of them and the applied effects of them. We will have to reconstruct the whole block again if we want to store them in the orphan pool.